package-system/mcpp/mcpp_2.7.2_az.patch

```diff
@@ -242,15 +242,68 @@ index c57eaef..909be89 100644
              openum = OP_1;
          }
          break;
-@@ -1722,6 +1735,7 @@ com_start:
-                     sp -= 2;
-                     while (*sp != '\n')     /* Until end of line    */
-                         mcpp_fputc( *sp++, OUT);
-+                    mcpp_fputc('\n', OUT);
+@@ -1654,6 +1667,28 @@ static char *   parse_line( void)
+     size_t      com_size;
+     int         c;
+ 
++    // By O3DE:
++    // ---- The Goal:
++    // Match the number of lines between the source AZSL files, before preprocessing,
++    // and the AZSL files after preprocessing. Line number matching is important because it helps
++    // providing accurate line numbers when reporting errors either by AZSLc or DXC.
++    // 
++    // ---- The Problem:
++    // MCPP adds a new line after each comment line that starts with '//'
++    // 
++    // ---- Solution:
++    // Do not add the extra line ending character '\n'.
++    // 
++    // ---- How:
++    // We count here the amount of leading spaces before the first
++    // non space character is detected.
++    // This is used to decide whether comments are going to be printed out or
++    // skipped altogether.
++    // If the current line only contains comments, preceded ONLY by whitespaces then
++    // will print the comment in the output WITHOUT line ending character. Otherwise
++    // the comment won't make it into the output.
++    int         NumLeadingSpaces = 0;
++
+     if ((sp = get_line( FALSE)) == NULL)    /* Next logical line    */
+         return  NULL;                       /* End of a file        */
+     if (in_asm) {                           /* In #asm block        */
+@@ -1668,9 +1703,12 @@ static char *   parse_line( void)
+ 
+     while (char_type[ c = *sp++ & UCHARMAX] & HSP) {
+         if (mcpp_mode != POST_STD)
++        { // O3DE
+             /* Preserve line top horizontal white spaces    */
+             /*      as they are for human-readability       */
+             *tp++ = c;
++            ++NumLeadingSpaces; //O3DE
++        } // O3DE
+         /* Else skip the line top spaces    */
+     }
+     sp--;
+@@ -1719,9 +1757,15 @@ com_start:
+                     cwarn( "Parsed \"//\" as comment"       /* _W2_ */
+                             , NULL, 0L, NULL);
+                 if (keep_comments) {
+-                    sp -= 2;
+-                    while (*sp != '\n')     /* Until end of line    */
+-                        mcpp_fputc( *sp++, OUT);
++                    const int numCharactersBeforeComment = (tp - temp);
++                    if (numCharactersBeforeComment == NumLeadingSpaces)
++                    {
++                        // O3DE: It is safe to print the comments
++                        sp -= 2;
++                        while (*sp != '\n')     /* Until end of line    */
```

Comment:
Silly question but, why doesn't changing just this line to `*(sp + 1) != '\n'` not work to address the issue described in the description?

Reply:
It's hard to do code review over the diff of a patch...
Maybe in the future we should move MCPP to GIT. Will need to follow up with the original author.

To see what's going on you'd have to see the original 'src/support.c' and that's why I added an //O3DE next to the lines I'm inserting. The original code is a massive pure 'c' state machine that is brittle and prone to bugs to any small change in pointer arithmetic. It took me a while to understand that state machine and believe me when I say that innocent or trivial changes to the pointer math they do always has collateral damage somewhere else ;-)

With this PR the goal is accomplished, not worth investing more time than necessary.

Reply:
Here is the original snippet:
```
                if (keep_comments) {
                    sp -= 2;
                    while (*sp != '\n')     /* Until end of line    */
                        mcpp_fputc( *sp++, OUT);
                    mcpp_fputc('\n', OUT);
                }
```

Reply:
I actually downloaded the original package when I was doing the review because it's really hard to review a diff/patch as you say. The original code looks like this:

```c
                if (keep_comments) {
                    sp -= 2;
                    while (*sp != '\n')     /* Until end of line    */
                        mcpp_fputc( *sp++, OUT);
                }
                goto  end_line;
```
It seems like all we want to do is strip a single `\n` from the output stream, so something like this right?
```c
                if (keep_comments) {
                    sp -= 2;
                    while (*(sp + 1) != '\n')     /* Until one character before the end of line    */
                        mcpp_fputc( *sp++, OUT);
                    ++sp; // Advance to original position
                }
                goto  end_line;
```

Reply:
Is the issue specifically things like carriage return characters maybe?

Reply:
No. You see it says:
`goto  end_line;`
At the start of 'goto end_line' there's another: